datafusion/src/optimizer/projection_push_down.rs

```diff
@@ -75,9 +78,12 @@ fn get_projected_schema(
     //
     // we discard non-existing columns because some column names are not part of the schema,
     // e.g. when the column derives from an aggregation
-    let mut projection: Vec<usize> = required_columns
+    //
+    // Use BTreeSet to remove potential duplicates (e.g. union) as
+    // well as to sort the projection to ensure deterministic behavior
+    let mut projection: BTreeSet<usize> = required_columns
         .iter()
-        .filter(|c| c.relation.as_ref() == table_name)
+        .filter(|c| c.relation.is_none() || c.relation.as_ref() == table_name)
```

Comment:
This is the core change -- don't compare the relation qualifier if there is none -- otherwise if `c = Column { relation: None, name: "a"}` and the table name is `Some("foo")` the column will be filtered, even if `foo` has a column named `a`

Reply:
good catch 👍